Added multiplicity related functions for projective plane curves and a…

delphinepol opened the pull request:
…n abstract type plane curve. Reordered the functions

Added multiplicity, intersection_multiplicity, aretransverse, homogenization for ProjPlaneCurve, reunited the functions hash, degree, curve_components, isirreducible, isreduced, reduction and union of AffinePlaneCurve and ProjPlaneCurve using the abstract type PlaneCurve.

delphinepol committed:
Added multiplicity related function for projective plane curves and an abstract type plane curve. Reordered the functions
Added multiplicity, intersection_multiplicity, aretransverse, homogenization for ProjPlaneCurve, reunited the functions hash, degree, curve_components, isirreducible, isreduced, reduction and union of AffinePlaneCurve and ProjPlaneCurve using the abstract type PlaneCurve.

rfourquet commented on examples/AffinePlaneCurve.jl:
```diff
@@ -213,12 +61,12 @@ end
      return Vector{AffinePlaneCurve}()
   else
-     return AffinePlaneCurve(G)
+     return Vector([AffinePlaneCurve(G)])
```
`[...]` is already a `Vector` constructor:
```suggestion
     return [AffinePlaneCurve(G)]
```

thofma commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+    rC.components = Dict(AffinePlaneCurve(D.eq) => 1 for D in keys(C.components))
+  elseif typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
+    rC = ProjPlaneCurve(F)
+    rC.components = Dict(ProjPlaneCurve(D.eq) => 1 for D in keys(C.components))
```
```suggestion
    rC.components = Dict(ProjPlaneCurve(D.eq.f) => 1 for D in keys(C.components))

```

fingolfin replied:
Another good reason for `defining_equation` which @thofma suggested: this kind of mistake is less likely to occur :-)

delphinepol replied:
Thanks for the comments. By forbidding the defining equation to be constant, I forbid also curves without components. Regarding your suggestion to write C.components = Dict(T(x) => e for (x, e) in D)),  it does not work, T(x) does not create a plane curve, it throws the error "ERROR: MethodError: no method matching ProjPlaneCurve{fmpq}(::fmpq_mpoly)". Is there a way to fix it ?

thofma replied:
Ah, I see. Yes, there is a way, but is a bit more intrusive, so I would ignore it for this PR here.

thofma reviewed (commented):
This looks good. A few small remarks.
1. There are lot of branches that look like

    ```julia
    if C.degree < 0 && typeof(C) == AffinePlaneCurve{typeof(lc(C.eq))}
      C.degree = total_degree(C.eq)
    elseif C.degree < 0 && typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
      C.degree = total_degree(C.eq.f)
    ```

    I think it would make sense to have

    ```julia
    defining_equation(C::AffinePlaneCurve) = C.eq
    defining_equation(C::ProjPlaneCurve) = C.eq.f
    ```

    Then one could write this as

    ```julia
    function Oscar.degree(C::PlaneCurve)
      if C.degree < 0
        C.degree = total_degree(defining_equation(C))
      end
      return(C.degree)
    end
    ```

    and also at other places.

    In julia it is best to avoid those `type` comparisons by refactoring the code as above.
2. The following

    ```julia
    function compo(C::PlaneCurve)
      if typeof(C) == AffinePlaneCurve{typeof(lc(C.eq))}
        D = factor(C.eq)
        C.components = Dict(AffinePlaneCurve(x) => D.fac[x] for x in keys(D.fac))
      elseif typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
        D = factor(C.eq.f)
        C.components = Dict(ProjPlaneCurve(x) => D.fac[x] for x in keys(D.fac))
      end
    end
    ```

    can be refactored to

    ```julia
    function compo(C::T) where T <: PlaneCurve
      D = factor(defining_equation(C))
      C.components = Dict(T(x) => e for (x, e) in D))
    end
    ```

    (Replaced `T(x) => D.fac[x] for x in keys(D.fac)` by `T(x) => e for (x,e) in D`. One can directly iterate over the factor object. Exploting that this is implemented by a dictionary is not future proof if we decide to change this.)
3. There are lot of blocks of the form

    ```julia
    if isempty(C.components)
      compo(C)
    end
    ```

    I would write an `_assure_has_components(C)` function which does this and just call this. 
4. It seems that `length(C.components) == 0` is used as a way to say the components have not been computed before? Do we allow curves with no components?

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+# The point is described by its coordinates.
+# ambient_dim gives the dimension of the space to which belongs the point.
+
+mutable struct Point{S}
```
Is `S` again meant to be the type of a field element? Then a type annotation saying so would be helpful:
```suggestion
mutable struct Point{S <: FieldElem}
```

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+# ambient_dim gives the dimension of the space to which belongs the point.
+
+mutable struct Point{S}
+  coord::Array{S, 1}
```
```suggestion
  coord::Vector{S}
```

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+
+mutable struct Point{S}
+  coord::Array{S, 1}
+  ambient_dim::Int64
```
Unless you really think that you'll need dimensions larger than 2^31, I'd use the "native" int type here and everywhere else instead:
```suggestion
  ambient_dim::Int
```

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+  eq::Oscar.MPolyElem{S}                  # Equation of the curve (polynomial in two variables)
+  degree::Int64                           # degree of the equation of the curve
+  dimension::Int64                        # dimension of the curve (as a variety)
+  components::Dict
```
Try to make the type concrete; my guess:
```suggestion
  components::Dict{AffinePlaneCurve{S}, Int}
```

thofma replied:
I don't think recursive definitions like this are currently supported. To avoid the type instability, we could have a `components(..)` function to access the components, which will have the explicit type annotation.

fingolfin replied:
It seems to work for `ProjPlaneCurve`, though?

thofma replied:
Oh, interesting, you are right.
```julia
struct A
  x::A
end
```
does not work, but
```julia
struct A
  x::Dict{A, Int}
end
```
works. I take everythinig back.

delphinepol replied:
It is indeed working.

fingolfin replied:
The former does not work because a non-empty struct type cannot contain another struct of the same, as then instances would require infinite storage.
But a struct type can contain *pointers* to other instances of the same struct type, and that's essentially what `Dict` does.

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+       error("The defining equation must belong to a ring with two variables")
+    elseif isconstant(eq)
+       error("The defining equation must be non constant")
+    end
```
You could use here and elsewhere this style of error checks (but the current one is also OK):
```suggestion
    nvars(parent(eq)) == 2 || error("The defining equation must belong to a ring with two variables")
    !isconstant(eq) || error("The defining equation must be non-constant")
```

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+  elseif C.degree < 0 && typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
+    C.degree = total_degree(C.eq.f)
+  end
+  return(C.degree)
```
```suggestion
  return C.degree

```

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+     return false
+  else
+     return all(isone, values(C.components))
+  end
```
```suggestion
  return length(C.components) == 1 && all(isone, values(C.components))
```

fingolfin reviewed (commented):
@thofma already left a bunch of excellent comments, thanks! Here are a few more suggestions

fingolfin commented on examples/PlaneCurve.jl:
```diff
@@ -0,0 +1,273 @@
+  elseif typeof(C) == AffinePlaneCurve{typeof(lc(C.eq))}
+     return AffinePlaneCurve(F)
+  end
+end
```
Could also do this:
```suggestion
function Base.union(C::T, D::T)  where T <: PlaneCurve
     return T(C.eq*D.eq)
end
```

delphinepol replied:
I am working on a fix for some type errors, I allowed decorated or non decorated polynomial as an input for ProjPlaneCurve for the current working version, but it leads to some polynomials not leaving in the same ring. I am writing now some functions for decorated polynomials to stay type consistent.
To @fingolfin: as I already answered to @thofma  in another question, T(C.eq) does not work, it throws me an error.

fingolfin replied:
I think the problem here is that `T` is not equal to `AffinePlaneCurve` or `ProjPlaneCurve` but rather to a concrete parametrization of those, i.e. `ProjPlaneCurve{S}` for a specific type `S`/ And the type of `C.eq*D.eq` does not match.

I think that hints at a (hopefully minor) design issue -- but I am sure it could be resolved; it doesn't have to be right now, though. In the meantime, this ought to work (or just leave this bit as it is, and we deal with it later -- perhaps @thofma or @fieker or me can have a look at your code once it is merged to figure this out):
```suggestion
function Base.union(C::T, D::T)  where T <: AffinePlaneCurve
     return AffinePlaneCurve(C.eq*D.eq)
end
function Base.union(C::T, D::T)  where T <: ProjPlaneCurve
     return ProjPlaneCurve(C.eq*D.eq)
end
```

fieker commented:
While I appreciate the lively discussion, what about merging? And then possibly further improving?

thofma commented:
The code is not working as is. Do you still want to merge it?

fieker commented:
On Fri, Dec 04, 2020 at 01:33:13AM -0800, thofma wrote:
> The code is not working as is. Do you still want to merge it?
> 
It is it not working, then no....
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/194#issuecomment-738677495

fingolfin commented:
@delphinepol of course the comments by Tommy and myself are meant constructive as always; heed those you think are reasonable, ignore the rest (and of course feel free to ask more questions). Let us know once this is ready :-)

delphinepol committed:
Included comments
Included comments on type comparison, made the type more consistent for equations of ProjPlaneCurve, fixed intersection_multiplicity for ProjPlaneCurve
